File: collect_app/src/main/java/org/odk/collect/android/formentry/FormIndexAnimationHandler.java
package org.odk.collect.android.formentry;

import androidx.annotation.Nullable;

import org.javarosa.core.model.FormIndex;

/**
 * Responsible for determining how a new "screen" in a form should be animated to based on
 * the {@link FormIndex}.
 */
public class FormIndexAnimationHandler {

    public enum Direction {
        FORWARDS,
        BACKWARDS
    }

    private final Listener listener;
    private FormIndex lastIndex;

    public FormIndexAnimationHandler(Listener listener) {
        this.listener = listener;
    }

    public void handle(@Nullable FormIndex index) {
        if (index == null) {
            return;
        }

        if (lastIndex == null) {
            listener.onScreenRefresh();
        } else {
            if (index.compareTo(lastIndex) > 0) {
                listener.onScreenChange(Direction.FORWARDS);
            } else if (index.compareTo(lastIndex) < 0) {
                listener.onScreenChange(Direction.BACKWARDS);
            } else {
                listener.onScreenRefresh();
            }
        }

        lastIndex = index;
    }

    /**
     * Can be used to update the handler on the starting index in situations
     * where {@link #handle(FormIndex)} isn't be called.
     */
    public void setLastIndex(FormIndex lastIndex) {
        this.lastIndex = lastIndex;
    }

    public interface Listener {
        void onScreenChange(Direction direction);

        void onScreenRefresh();
    }
}

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormIndexAnimationHandler.kt
package org.odk.collect.android.formentry

import org.javarosa.core.model.FormIndex

/**
 * Responsible for determining how a new "screen" in a form should be animated to based on
 * the [FormIndex].
 */
class FormIndexAnimationHandler(private val listener: Listener) {

    enum class Direction {
        FORWARDS, BACKWARDS
    }

    private var lastIndex: FormIndex? = null

    fun handle(index: FormIndex) {
        if (lastIndex == null) {
            listener.onScreenRefresh(true)
        } else {
            if (index > lastIndex) {
                listener.onScreenChange(Direction.FORWARDS)
            } else if (index < lastIndex) {
                listener.onScreenChange(Direction.BACKWARDS)
            } else {
                listener.onScreenRefresh(false)
            }
        }

        lastIndex = index
    }

    /**
     * Can be used to update the handler on the starting index in situations
     * where [.handle] isn't be called.
     */
    fun setLastIndex(lastIndex: FormIndex?) {
        this.lastIndex = lastIndex
    }

    interface Listener {
        fun onScreenChange(direction: Direction?)
        fun onScreenRefresh(isFormStart: Boolean)
    }
}
